[DI] Allow explicit executor for torchscript interpreter

garroud opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46444 [DI] Allow explicit executor for torchscript interpreter**

add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

Commit:
[DI] Allow explicit executor for torchscript interpreter

add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

dzhulgakov commented on test/cpp/jit/test_interpreter.cpp:
```diff
@@ -138,5 +142,38 @@ TEST(InterpreterTest, Basic_CUDA) {
+  demo = DemoModule()
+  torch.jit.save(torch.jit.script(demo), 'test_interpreter_async.pth')
+  */
+  auto model = load("caffe2/test/cpp/jit/test_interpreter_async.pth");
```
this won't work in OSS CI - you need to change it to a relative path (or wait for the parser problem to be fixed)

examples:
https://github.com/pytorch/pytorch/blob/master/test/jit/test_save_load.py#L13
https://github.com/pytorch/pytorch/blob/master/test/quantization/test_backward_compatibility.py#L29

houseroad replied:
I thought we are able to load the models by copying the generated code, so we are still not able to do so?

garroud replied:
forks seems to have problem with the parser.

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

garroud commented:
Do we allow to use Boost here? the build failure seems due to I did not add Boost into Cmake

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

houseroad reviewed (commented):
Boost is not dependency of PyTorch, and I don't think it's due to Boost ( I didn't see any dependency on Boost).

Related error message: 
```
test_interpreter.cpp:(.text+0x2b7e): undefined reference to `torch::jit::InterpreterState::runAsync(std::vector<c10::IValue, std::allocator<c10::IValue> >&)'
```
You probably need to expose `InterpreterState` as well.

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

houseroad reviewed (commented):
Test failures in backward compatibility check is unrelated, please rebase your diff and reexport the diff to github.

pytorch_linux_bionic_py3_6_clang9_test's failure is related, seems the model fails to load.

houseroad reviewed (commented):
Also please use .pt extension instead .pth.

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

garroud committed:
add test module externally

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

garroud committed:
add test module

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

garroud committed:
add test module

Commit:
Update on "[DI] Allow explicit executor for torchscript interpreter"


add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

[ghstack-poisoned]

garroud commented:
hgexport not work with binary file. close this PR.
